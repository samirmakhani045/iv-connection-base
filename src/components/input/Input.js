import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { isIconValid } from '@utils/stories/getAllIcons';

import { Wrapper, StyledInput, Label, Span, StyledIcon, Error } from './Input.styles';

const Input = (props) => {
  const { disabled, id, label, type, name, value, icon, error, required, ...remainingProps } =
    props;
  const [inputValue, setInputValue] = useState(value);

  const handleOnChange = (ev) => {
    setInputValue(ev.target.value);
  };

  return (
    <Wrapper data-testid="wrapper">
      <Label id={`${id}-label`} htmlFor={id}>
        <StyledInput
          data-testid="input"
          id={id}
          name={name}
          type={type}
          disabled={disabled}
          required={required}
          value={inputValue}
          onChange={handleOnChange}
          withError={!!error}
          withIcon={!!isIconValid(icon)}
          aria-invalid={!!error}
          aria-required={required}
          {...(label && { 'aria-labelledby': `${id}-label` })}
          {...(error && { 'aria-describedby': `${id}-error` })}
          {...remainingProps}
        />
        {label && (
          <Span
            data-testid="inputLabel"
            withError={!!error}
            withIcon={!!isIconValid(icon)}
            withValue={!!inputValue}
          >
            {label}
          </Span>
        )}
        <StyledIcon data-testid="inputIcon" name={icon} size={12} withError={!!error} />
      </Label>
      {error && (
        <Error data-testid="inputError" id={`${id}-error`}>
          {error}
        </Error>
      )}
    </Wrapper>
  );
};

Input.propTypes = {
  /** Disables the input */
  disabled: PropTypes.bool,
  /** IDD of the input that is used on forms */
  id: PropTypes.string.isRequired,
  /** Label to describe the input */
  label: PropTypes.string.isRequired,
  /** Type of input control */
  type: PropTypes.oneOf(['text', 'number']),
  /** Name of the input that is used on forms */
  name: PropTypes.string,
  /** Value to fill the input */
  value: PropTypes.string,
  /** If set, icon that will be placed on the right side of the input */
  icon: PropTypes.string,
  /** Validation error */
  error: PropTypes.string,
  /** Set input as required */
  required: PropTypes.bool,
};

Input.defaultProps = {
  type: 'text',
  value: '',
};

export default Input;
