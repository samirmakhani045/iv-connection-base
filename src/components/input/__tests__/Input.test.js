import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent, { specialChars } from '@testing-library/user-event';

import Input from '../Input';

const defaultProps = {
  id: 'test',
  label: 'Label',
};

const Test = (props) => <Input {...defaultProps} {...props} />;

describe('<Input />', () => {
  it('should render correctly', () => {
    const { container } = render(<Test />);

    expect(screen.queryByTestId('wrapper')).toBeInTheDocument();
    expect(screen.queryByTestId('input')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a label, value and icon', () => {
    const { container } = render(<Test value="test" icon="Search" />);

    expect(screen.queryByTestId('inputLabel')).toBeInTheDocument();
    expect(screen.queryByTestId('inputIcon')).toBeInTheDocument();
    expect(screen.queryByTestId('input')).toHaveValue('test');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should handle a value change', () => {
    render(<Test value="this is a test" />);

    const input = screen.queryByTestId('input');

    input.setSelectionRange(10, 14);
    userEvent.type(input, `${specialChars.backspace}test1`);
    expect(input).toHaveValue('this is a test1');
  });

  it('should render a disabled input', () => {
    render(<Test value="test" disabled />);

    const input = screen.queryByTestId('input');

    userEvent.type(input, 'test1');
    expect(input).toHaveValue('test');
    expect(input).toHaveProperty('disabled');
  });

  it('should render a required input', () => {
    render(<Test required />);

    expect(screen.queryByTestId('input')).toHaveProperty('required');
  });

  it('should display an error message', () => {
    const { container } = render(<Test error="Error" />);

    expect(screen.queryByTestId('inputError')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
