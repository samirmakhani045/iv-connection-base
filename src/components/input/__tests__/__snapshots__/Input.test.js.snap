// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should display an error message 1`] = `
<div
  class="Inputstyles__Wrapper-sc-1mcot6p-0 MKIHn"
  data-testid="wrapper"
>
  <label
    class="Inputstyles__Label-sc-1mcot6p-1 cGGdri"
    for="test"
    id="test-label"
  >
    <input
      aria-describedby="test-error"
      aria-invalid="true"
      aria-labelledby="test-label"
      class="Inputstyles__StyledInput-sc-1mcot6p-4 hpdgNO"
      data-testid="input"
      id="test"
      type="text"
      value=""
    />
    <span
      class="Inputstyles__Span-sc-1mcot6p-2 iywYCO"
      data-testid="inputLabel"
    >
      Label
    </span>
  </label>
  <div
    class="Inputstyles__Error-sc-1mcot6p-5 ifvZfE"
    data-testid="inputError"
    id="test-error"
  >
    Error
  </div>
</div>
`;

exports[`<Input /> should render a label, value and icon 1`] = `
<div
  class="Inputstyles__Wrapper-sc-1mcot6p-0 MKIHn"
  data-testid="wrapper"
>
  <label
    class="Inputstyles__Label-sc-1mcot6p-1 cGGdri"
    for="test"
    id="test-label"
  >
    <input
      aria-invalid="false"
      aria-labelledby="test-label"
      class="Inputstyles__StyledInput-sc-1mcot6p-4 jvlSCZ"
      data-testid="input"
      id="test"
      type="text"
      value="test"
    />
    <span
      class="Inputstyles__Span-sc-1mcot6p-2 ktnSuP"
      data-testid="inputLabel"
    >
      Label
    </span>
    <div
      class="Iconstyles__IconWrapper-sc-1ghvgbo-0 jMuoDZ Inputstyles__StyledIcon-sc-1mcot6p-3 kXuUVb"
      data-testid="inputIcon"
      size="12"
    >
      <svg
        data-testid="icon"
        fill="#313336"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M15.5 14H14.71L14.43 13.73C15.41 12.59 16 11.11 16 9.5C16 5.91 13.09 3 9.5 3C5.91 3 3 5.91 3 9.5C3 13.09 5.91 16 9.5 16C11.11 16 12.59 15.41 13.73 14.43L14 14.71V15.5L19 20.49L20.49 19L15.5 14ZM9.5 14C7.01 14 5 11.99 5 9.5C5 7.01 7.01 5 9.5 5C11.99 5 14 7.01 14 9.5C14 11.99 11.99 14 9.5 14Z"
        />
      </svg>
    </div>
  </label>
</div>
`;

exports[`<Input /> should render correctly 1`] = `
<div
  class="Inputstyles__Wrapper-sc-1mcot6p-0 MKIHn"
  data-testid="wrapper"
>
  <label
    class="Inputstyles__Label-sc-1mcot6p-1 cGGdri"
    for="test"
    id="test-label"
  >
    <input
      aria-invalid="false"
      aria-labelledby="test-label"
      class="Inputstyles__StyledInput-sc-1mcot6p-4 gXtDrl"
      data-testid="input"
      id="test"
      type="text"
      value=""
    />
    <span
      class="Inputstyles__Span-sc-1mcot6p-2 eXpXFQ"
      data-testid="inputLabel"
    >
      Label
    </span>
  </label>
</div>
`;
